### DISCUSSION TITLE
Polymorphism and refs types

### DISCUSSION THREAD

**Discussion by DaniGuardiola** [radix-ui/primitives#665] [Help] [answered]
I'm having trouble correctly typing something.

Here's an ultra-minimal repro: https://codesandbox.io/s/mystifying-jennings-glrz5?file=/src/App.tsx

Basically, I want to dynamically set the element (`as` prop) based on some internal runtime state. I also need to pass it a ref. Because I cannot statically infer (to my best knowledge) what element it will be, I'm typing the ref as `HTMLElement`. This throws type errors because apparently it needs more specificity (`HTMLElement` not assignable to `HTMLDivElement`, etc).

Findings:

- If I only use `as`, it works.
- If I only use `ref`, it works.
- The conditional doesn't change the error (`as="button"` still throws an error)

What am I doing wrong here? I've tried many things, but none have worked so far. Thanks in advance for any help! :smile:

**Comment by jjenzz** [comment, marked as answer]
This looks like it could be a bug as I would have expected you would be able to do the following which also doesn't work:

```jsx
const ref = React.useRef<HTMLButtonElement | HTMLDivElement>(null);
```

It appears this [should be possible](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAIKoAKxYqAPJzhYAHjCwA7ACboA1liYQCcAFIBlABoA6AJISYUYBNTBcAUQA2WEJJioAfHAC8cAN4pUALg6oANHEoCb2w8GC1sAn4Q-C1La1sIwQdHAF86AgBXCXxgCAlMCAgYJOExSRk4eUVldW09AyMTcysbfQcACjBeLx8eCD4k+wBKNzS6XDzUeGBUeqwoUOAANzYXAwysegnjeEDnTCowjNQsRIAJABUAWQAZACEMmBg8uNb4AB84S9uAEWXX2z2doSDIWCxDeh0DCFGDtdxobwzOYLHIrOAAfnIACNHs8JGQ4N4yFJlmR-HsUkMgA) so I'll raise an issue from this discussion for us to look into further.

EDIT: I've raised the issue [here](https://github.com/radix-ui/primitives/issues/675) if you'd like to follow progress.

**Reply by DaniGuardiola** [reply]
Thanks :)
